File: frontend/src/domain/status/status.ts
export type Status = {
	code: number;
	message: string;
};

export const STATUS = {
	// common status
	OK: { code: 0, message: "ok" },
	ERR_UNKNOWN: { code: 1, message: "unknown error" },

	// auth domain status
	ERR_AUTH_INVALID_VERIFICATION_TOKEN: {
		code: 1000,
		message: "invalid verification token",
	},

	// email address domain status
	ERR_EMAIL_ADDRESS_INVALID_EMAIL_FORMAT: {
		code: 2000,
		message: "invalid email format",
	},

	// payment category domain status
	ERR_PAYMENT_CATEGORY_INVALID_COLOR_NAME: {
		code: 3000,
		message: "invalid color name",
	},
	ERR_PAYMENT_CATEGORY_INVALID_COLOR_TOME: {
		code: 3001,
		message: "invalid color tone",
	},
} as const satisfies Record<string, Status>;

export function getStatusByCode(code: number | undefined): Status {
	const status = Object.values(STATUS).find((status) => status.code === code);
	if (!status) {
		return STATUS.ERR_UNKNOWN;
	}
	return status;
}

File: backend/domain/status/status.go
package status

import "fmt"

type Status struct {
	Code    int
	Message string
}

func NewStatus(code int, message string) *Status {
	return &Status{
		Code:    code,
		Message: message,
	}
}

func (e *Status) Error() string {
	return fmt.Sprintf("Status Code %d: %s", e.Code, e.Message)
}

var (
	// common status
	Ok         = NewStatus(0, "ok")
	ErrUnknown = NewStatus(1, "unknown error")

	// auth domain status
	ErrAuthInvalidVerificationToken = NewStatus(1000, "invalid verification token")

	// email address domain status
	ErrEmailAddressInvalidEmailFormat = NewStatus(2000, "invalid email format")

	// payment category domain status
	ErrPaymentCategoryInvalidColorName = NewStatus(3000, "invalid color name")
	ErrPaymentCategoryInvalidColorTone = NewStatus(3001, "invalid color tone")
)
